File: src/lib/physics/hawking.ts
/**
 * Hawking radiation primitives (visualization, illustrative).
 *
 * Mirrors the Rust gravitas::quantum::hawking surface for use in
 * React components. The blackbody Planck shape is the standard
 * pedagogical curve; greybody transmission factors are out of scope
 * per ADR-0026 honest-labeling rule.
 *
 * References:
 * - Hawking 1974, Nature 248, 30.
 * - Hawking 1975, Commun. Math. Phys. 43, 199.
 * - Wien 1893 (modern displacement constant 2.898e-3 m·K → 5.879e10 Hz/K).
 */

const SI_C = 299_792_458;
const SI_KB = 1.380649e-23;
const SI_HBAR = 1.054_571_817e-34;
const SI_G = 6.6743e-11;
const WIEN_FREQUENCY_CONSTANT_HZ_PER_K = 5.878_925_757e10;

/**
 * Hawking temperature for a Kerr hole of given SI rest mass and
 * dimensionless spin a* in [-1, 1]. Returns 0 at a* = ±1 (extremal
 * limit: surface gravity vanishes).
 */
export function hawkingTemperatureKelvin(
  massKg: number,
  spinStar: number,
): number {
  if (massKg <= 0) return 0;
  const aStar = Math.max(-1, Math.min(1, spinStar));
  const disc = Math.sqrt(Math.max(0, 1 - aStar * aStar));
  const rPlus = 1 + disc;
  const rMinus = 1 - disc;
  // Geometric surface gravity κ; SI conversion below picks up M and the
  // units of the gravitational constant.
  const kappaGeom = (rPlus - rMinus) / (2 * (rPlus * rPlus + aStar * aStar));
  const kappaSi = (kappaGeom * SI_C ** 3) / (SI_G * massKg);
  return (SI_HBAR * kappaSi) / (2 * Math.PI * SI_KB);
}

/**
 * Wien displacement law: peak frequency of the Planck spectrum at T.
 */
export function wienPeakFrequency(temperatureK: number): number {
  return WIEN_FREQUENCY_CONSTANT_HZ_PER_K * Math.max(0, temperatureK);
}

/**
 * Planck spectral radiance B_ν(T) in SI W·m⁻²·Hz⁻¹·sr⁻¹. Returns 0
 * for non-positive inputs and for h ν / k T > 700 (where the
 * exponential overflows IEEE 754).
 */
export function hawkingSpectrumPlanck(
  freqHz: number,
  temperatureK: number,
): number {
  if (freqHz <= 0 || temperatureK <= 0) return 0;
  const h = SI_HBAR * 2 * Math.PI;
  const exponent = (h * freqHz) / (SI_KB * temperatureK);
  if (exponent > 700) return 0;
  const prefactor = (2 * h * freqHz ** 3) / (SI_C * SI_C);
  return prefactor / Math.expm1(exponent);
}

/**
 * Sample the Hawking spectrum on a logarithmically-spaced frequency
 * grid centred on the Wien peak. Returns one (frequency, radiance)
 * pair per sample, suitable for a recharts log-log plot.
 */
export function sampleHawkingSpectrum(
  temperatureK: number,
  decadesEachSide = 3,
  samples = 64,
): Array<{ freqHz: number; radiance: number }> {
  if (temperatureK <= 0 || samples < 2) return [];
  const peak = wienPeakFrequency(temperatureK);
  if (peak <= 0) return [];
  const log10Peak = Math.log10(peak);
  const out: Array<{ freqHz: number; radiance: number }> = [];
  for (let i = 0; i < samples; i++) {
    const t = i / (samples - 1);
    const log10Freq = log10Peak + (2 * t - 1) * decadesEachSide;
    const freqHz = 10 ** log10Freq;
    out.push({ freqHz, radiance: hawkingSpectrumPlanck(freqHz, temperatureK) });
  }
  return out;
}

File: physics-engine/gravitas-core/src/quantum/hawking.rs
//! Hawking radiation temperature.

use crate::constants::{SI_C, SI_HBAR, SI_KB};

/// Hawking temperature for a Kerr black hole.
///
/// T_H = (hbar * c^3) / (8 pi G M k_B) * (r+ - r-) / (r+^2 + a^2)
///
/// For the Schwarzschild limit (a=0): T_H = hbar c^3 / (8 pi G M k_B)
///
/// # Arguments
/// - `mass_kg` -- Mass in kilograms (SI)
/// - `spin` -- Dimensionless spin a* in [-1, 1]
pub fn hawking_temperature(mass_kg: f64, spin: f64) -> f64 {
    let g_si = 6.674_30e-11;
    let a_star = spin.clamp(-1.0, 1.0);

    // In geometric units: M = 1, a = a*
    // r_+ = 1 + sqrt(1 - a*^2),  r_- = 1 - sqrt(1 - a*^2)
    let disc = (1.0 - a_star * a_star).max(0.0).sqrt();
    let r_plus = 1.0 + disc;
    let r_minus = 1.0 - disc;

    // Surface gravity kappa = (r+ - r-) / (2 * (r+^2 + a*^2))
    let kappa = (r_plus - r_minus) / (2.0 * (r_plus * r_plus + a_star * a_star));

    // T_H = hbar * kappa_SI / (2 pi k_B)
    // kappa_SI = kappa * c^3 / (G * M)
    let kappa_si = kappa * SI_C * SI_C * SI_C / (g_si * mass_kg);

    SI_HBAR * kappa_si / (2.0 * std::f64::consts::PI * SI_KB)
}

/// Spectral radiance of a Hawking-emitting hole at temperature T_H,
/// modelled as a blackbody (no greybody factor; rule 12 §"no false
/// claims" applies — the visualization panel uses this as a
/// Planck-shape illustrative spectrum, not the full grey-body
/// transmission coefficient).
///
/// Returns B_ν(T) in W·m⁻²·Hz⁻¹·sr⁻¹ per the SI Planck law:
///
///   B_ν(T) = (2 h ν³ / c²) / (exp(h ν / k_B T) − 1).
///
/// Caller passes frequency in Hz and the Hawking temperature in K
/// (computed via `hawking_temperature` for a given M, a*).
#[must_use]
pub fn hawking_spectrum_planck(freq_hz: f64, temperature_k: f64) -> f64 {
    if temperature_k <= 0.0 || freq_hz <= 0.0 {
        return 0.0;
    }
    let h = SI_HBAR * 2.0 * std::f64::consts::PI;
    let exponent = h * freq_hz / (SI_KB * temperature_k);
    if exponent > 700.0 {
        // exp would overflow; the Planck law is exponentially small,
        // so return zero rather than a NaN from inf-inf.
        return 0.0;
    }
    let prefactor = 2.0 * h * freq_hz.powi(3) / (SI_C * SI_C);
    prefactor / exponent.exp_m1()
}

/// Wien displacement law for a Hawking-emitting hole: peak frequency
/// of the Planck spectrum at a given temperature.
///
///   ν_peak / T = 5.879 × 10¹⁰ Hz/K   (Wien 1893, modern value).
///
/// Useful for the visualization panel's x-axis tick: the peak
/// frequency tells the operator which band the hole shines in.
#[must_use]
pub fn wien_peak_frequency(temperature_k: f64) -> f64 {
    const WIEN_FREQUENCY_CONSTANT_HZ_PER_K: f64 = 5.878_925_757e10;
    WIEN_FREQUENCY_CONSTANT_HZ_PER_K * temperature_k.max(0.0)
}
